qiime2/sdk/usage.py

```diff
@@ -8,60 +8,194 @@
 
 import abc
 import re
-import types
+from types import MappingProxyType
+from typing import Tuple, Callable, Optional, Union, List, Set, Dict
 
 from qiime2 import sdk, metadata
+from qiime2.core.type import MethodSignature, PipelineSignature
 
-# TODO: docstrings
+# Is Metadata sufficient for typing, or should we specify MetadataColumn
+# as well?
+Factory = Callable[..., Union[metadata.Metadata, sdk.Artifact]]
```

Comment:
Does python not blow up on this ellipsis??!?!

Reply:
Lolz, neato, eh?

>`Callable[..., ReturnType]` (literal ellipsis) can be used to type hint a callable taking any number of arguments and returning ReturnType.

https://docs.python.org/3.7/library/typing.html#typing.Callable